### DISCUSSION TITLE
Supporting keyset pagination via row-wise multi-column comparisons

### DISCUSSION THREAD

**Discussion by nhoughto** [micronaut-projects/micronaut-data#1108] [General]
We are looking at trying to build on micronaut-data and extending the AOT codegen behaviours to reduce a bunch of our boilerplate around custom pagination, ordering, filtering etc in our project. We expect to do this via tweaking the compile time AOT data-processor via ServiceLoaders.

Seems like we should be able to do this with the extension points that already exist, this was more a question on the projects position on this kind of thing? Is there a plan to support this kind of thing in the future via a more out of the box approach?

**Comment by dstepanov** [comment]
Depends on what do you want to do, maybe you can describe it in more detail.
You can do a lot with a simple method interceptor https://docs.micronaut.io/latest/guide/#aroundAdvice and position it before or after the data interceptor.

**Reply by nhoughto** [reply]
ah cool, discussions? didn't even know that was a thing!

So the number one thing we want to do is support keyset pagination via row-wise comparison in a `WHERE` clause, at the moment this doesn't seem possible. Essentially we need to use row-wise comparison (https://www.postgresql.org/docs/9.2/functions-comparisons.html) instead of a standard comparison for some queries, based on argument types i suppose, simila r to how arrays are handled maybe.

Since row-wise comparison isn't part of the SQL standard I thought it might be tricky to have vendor specific stuff, so thought that maybe an extension rather than an upstream feature would be a better fit. Obviously you guys know better though.

**Reply by dstepanov** [reply]
If you are using Micronaut Data JDBC/R2dbc we generate queries for the specific databases so it's ok to have vendor-specific stuff, there are already some differences in how arrays are generated.
Unfortunately how queries are stored is mostly internally specified and might change in the future, so I would suggest trying to replicate it.
The best would be to contribute the feature, if you have something specific we can talk about the best way to do it.

**Reply by nhoughto** [reply]
Looking at the processor code, my first thought would be to extend `io.micronaut.data.model.query.Criteria` to add some new row-wise operators? `rowwiseGt` `rowwiseLt` `rowwiseEq` etc which would then capture the context of forming the WHERE clause as `.. WHERE (col1, col2) > (val1, val2)` (row wise comparison)  instead of `WHERE col1=val1 AND col2=val2`?

Seems like the tricky part would be passing 1 argument that represents N columns?

Naively my Repository method would be something like:

`List<Thing> findByAccountEqualsOrderByNameAndDate(int account, KeySetPageable cursor)`

Building on Pagination?  So the keyset pagination isn't part of the method name / regular criteria same as normal Pagination, it is contained in the KeySetPageable, and KeySetPageable could contain 1-N columns and 1-N  matching values that would form up the `.. WHERE ROW(col1, col2) > (val1, val2)` bit.

Not really sure the knock-on effects of this, expect some changes to processor to persist the metadata, then runtime to execute it.

I've revisited our needs for extending micronaut-data, and whilst we have many needs to reduce boiler-plate etc, the current pain point it seems is around a lack of keyset pagination forcing use of `@Query` escape hatch, that then has a bunch of other knock-on effects like losing proper query projections on joins etc. So i think this is really a question about keyset pagination and not about extending anymore.. i'll change the discussion topic.

**Reply by nhoughto** [reply]
Also it seems like row-wise operation / row constructor comparison like `.. WHERE (col1, col2) > (val1, val2)` have support in most/some databases, with some edgecases around null handling that shouldn't be an issue here. So maybe the 'its too DB specific' thought i had wasn't valid to start with.

**Reply by dstepanov** [reply]
BTW Why do you need row-wise for the keyset pagination? Can you just implement it without? What is the benefit of having row-wise syntax?

**Reply by nhoughto** [reply]
maybe im missing something, but I thought multi column keyset pagination requires it? i need to specify a cursor with col1=a,col2=1234 and i want to specify to only return rows `>` both those columns from an ordered set. Can you do that without row-wise comparisons? 🤔 

`WHERE col1>a AND col2>1234` isnt the same as `WHERE (col1, col2) > (a,1234)`

**Reply by nhoughto** [reply]
Looking at the code more, it seems like what I want is similar to Pageable/Slice but quite different, those things make assumptions about the mechanics of how the Page is retrieved, it basically assumes `OFFSET .. LIMIT ..` and requires pageIndex and pageSize, where as keyset pagination will have a LIMIT but no OFFSET, and will have a `WHERE (col1, col2) > (cursorValue1, cursorValue2)` instead.

So it seems like a proper implementation for this would be quite invasive? As you would likely want to redefine the current Pagination as NumberedPagination or something and abstract the method of paginating with pageIndex+pageSize being one method and keyset being another, otherwise i'd effectively duplicate `Pageable` support everywhere and call it `KeySetPageable` which doesn't feel great?

**Reply by dstepanov** [reply]
I haven't seen row-wise being used so much so I think it's just a syntax sugar and it can be written in a different way. I thing ‘KeySetPageable’ is a way to go, if I'm correct you would have to define which column is actually the key, right? And this kind of pagination doesn't support sorting for that you would have to choose a different key-column.

**Reply by nhoughto** [reply]
I fixed this without changing the data-processor portion by using RepositoryOperations.findStream() directly and subclassing a few bits. So can close this i suppose.
